File: frontend/src/types/dsa.ts
/**
 * Types for individual DSA problems.
 *
 * These mirror the payloads served by the DSA service, which are generated
 * from the Python files under backend/services/dsa/content by extract.py.
 */

/** A problem is pending when its source file exists but has no solution yet. */
export type ProblemStatus = 'solved' | 'pending';

export interface ProblemExample {
  input: string;
  output: string;
  explanation?: string;
}

export interface ProblemComplexity {
  time?: string;
  space?: string;
}

/** One implementation. A file often holds several named variants. */
export interface ProblemSolution {
  language: string;
  code: string;
  name?: string;
  complexity?: ProblemComplexity;
}

export interface DSAProblem {
  id: string;
  topicId: string;
  title: string;
  /** Folder titles between the topic and the file, outermost first. */
  sectionPath: string[];
  statement: string;
  constraints?: string;
  /** Trailing write-up the author left below the code. */
  notes?: string;
  examples: ProblemExample[];
  solutions: ProblemSolution[];
  difficulty?: string;
  status: ProblemStatus;
  sourceFile: string;
  sortKey: string;
}

/** List payload; omits statements and solution bodies to keep responses small. */
export interface DSAProblemSummary {
  id: string;
  topicId: string;
  title: string;
  sectionPath: string[];
  difficulty?: string;
  status: ProblemStatus;
  sortKey: string;
  hasSolution: boolean;
  hasStatement: boolean;
  solutionCount: number;
}

/** A problem is worth opening when there is something to read or run. */
export function isProblemOpenable(problem: DSAProblemSummary): boolean {
  return problem.hasSolution || problem.hasStatement;
}

/** Groups problems into their sections while preserving curriculum order. */
export function groupBySection(
  problems: DSAProblemSummary[]
): Array<{ section: string; problems: DSAProblemSummary[] }> {
  const groups: Array<{ section: string; problems: DSAProblemSummary[] }> = [];
  const index = new Map<string, number>();

  for (const problem of problems) {
    const section = problem.sectionPath.join(' / ') || 'Overview';
    const existing = index.get(section);
    if (existing === undefined) {
      index.set(section, groups.length);
      groups.push({ section, problems: [problem] });
    } else {
      groups[existing].problems.push(problem);
    }
  }

  return groups;
}

File: backend/services/dsa/models/dsa.go
package models

import "time"

// DSATopic represents a Data Structures & Algorithms topic
type DSATopic struct {
	ID            string     `json:"id" db:"id"`
	Title         string     `json:"title" db:"title"`
	Category      string     `json:"category" db:"category"`
	Description   string     `json:"description" db:"description"`
	Difficulty    string     `json:"difficulty" db:"difficulty"` // Easy, Medium, Hard
	Progress      int        `json:"progress" db:"progress"`
	Icon          string     `json:"icon" db:"icon"`
	Color         string     `json:"color" db:"color"`
	FolderPath    string     `json:"folderPath" db:"folder_path"`
	ProblemCount  int        `json:"problemCount" db:"problem_count"`
	SolvedCount   int        `json:"solvedCount" db:"solved_count"`
	Subcomponents []string   `json:"subcomponents,omitempty"` // Stored as JSON in DB
	Content       DSAContent `json:"content"`                 // Stored as JSON in DB
	CreatedAt     time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt     time.Time  `json:"updated_at" db:"updated_at"`
}

// DSAContent represents the detailed content of a DSA topic
type DSAContent struct {
	Explanation   string        `json:"explanation"`
	Examples      []DSAExample  `json:"examples"`
	CodeTemplates CodeTemplates `json:"codeTemplates"`
	TestCases     []TestCase    `json:"testCases"`
	Hints         []string      `json:"hints"`
	Solution      string        `json:"solution"`
}

// DSAExample represents an example for a DSA problem
type DSAExample struct {
	Input       string `json:"input"`
	Output      string `json:"output"`
	Explanation string `json:"explanation"`
}

// CodeTemplates represents code templates in different languages
type CodeTemplates struct {
	JavaScript string `json:"javascript"`
	Python     string `json:"python"`
	Java       string `json:"java"`
	CPP        string `json:"cpp"`
}

// TestCase represents a test case for a DSA problem
type TestCase struct {
	Input          string `json:"input"`
	ExpectedOutput string `json:"expectedOutput"`
}

// DSACategory represents a DSA category
type DSACategory struct {
	Name  string `json:"name"`
	Count int    `json:"count"`
}

// Problem status values. A problem is pending when its source file exists in
// the curriculum but does not contain a solution yet.
const (
	ProblemStatusSolved  = "solved"
	ProblemStatusPending = "pending"
)

// Complexity holds the time and space analysis for one solution variant.
type Complexity struct {
	Time  string `json:"time,omitempty"`
	Space string `json:"space,omitempty"`
}

// DSASolution is one implementation of a problem. A single file often contains
// several named variants, for example a hash map approach and a sorting one.
type DSASolution struct {
	Language   string      `json:"language"`
	Code       string      `json:"code"`
	Name       string      `json:"name,omitempty"`
	Complexity *Complexity `json:"complexity,omitempty"`
}

// DSAProblem is a single problem from the content tree.
type DSAProblem struct {
	ID          string        `json:"id" db:"id"`
	TopicID     string        `json:"topicId" db:"topic_id"`
	Title       string        `json:"title" db:"title"`
	SectionPath []string      `json:"sectionPath"`
	Statement   string        `json:"statement"`
	Constraints string        `json:"constraints,omitempty"`
	Notes       string        `json:"notes,omitempty"`
	Examples    []DSAExample  `json:"examples"`
	Solutions   []DSASolution `json:"solutions"`
	Difficulty  string        `json:"difficulty,omitempty"`
	Status      string        `json:"status"`
	SourceFile  string        `json:"sourceFile"`
	SortKey     string        `json:"sortKey"`
}

// DSAProblemSummary is the list payload. It deliberately omits statements and
// solution bodies, which would otherwise make a topic like Graphs, with 149
// problems, an enormous response.
type DSAProblemSummary struct {
	ID            string   `json:"id"`
	TopicID       string   `json:"topicId"`
	Title         string   `json:"title"`
	SectionPath   []string `json:"sectionPath"`
	Difficulty    string   `json:"difficulty,omitempty"`
	Status        string   `json:"status"`
	SortKey       string   `json:"sortKey"`
	HasSolution   bool     `json:"hasSolution"`
	HasStatement  bool     `json:"hasStatement"`
	SolutionCount int      `json:"solutionCount"`
}
